File: go/ch09/beam_fill.go
// go/ch09/beam_fill.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const beamWidth = 80

type dir struct {
	dr, dc int
	ch     byte
}

var dirs = []dir{{-1, 0, 'U'}, {1, 0, 'D'}, {0, -1, 'L'}, {0, 1, 'R'}}

type beamNode struct {
	priority int
	depth    int
	r, c     int
	visited  uint64
	score    int
	path     string
	index    int
}

type beamHeap []*beamNode

func (h beamHeap) Len() int           { return len(h) }
func (h beamHeap) Less(i, j int) bool { return h[i].priority < h[j].priority }
func (h beamHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
	h[i].index = i
	h[j].index = j
}
func (h *beamHeap) Push(x interface{}) {
	n := len(*h)
	node := x.(*beamNode)
	node.index = n
	*h = append(*h, node)
}
func (h *beamHeap) Pop() interface{} {
	old := *h
	n := len(old)
	node := old[n-1]
	old[n-1] = nil
	node.index = -1
	*h = old[:n-1]
	return node
}

func readGrid(sc *bufio.Scanner) (int, [][]int) {
	sc.Scan()
	n, _ := strconv.Atoi(sc.Text())
	grid := make([][]int, n)
	for i := 0; i < n; i++ {
		sc.Scan()
		parts := strings.Fields(sc.Text())
		grid[i] = make([]int, n)
		for j, p := range parts {
			grid[i][j], _ = strconv.Atoi(p)
		}
	}
	return n, grid
}

func cellID(r, c, n int) int {
	return r*n + c
}

func heuristic(grid [][]int, n int, visited uint64, r, c int) int {
	est := 0
	for _, d := range dirs {
		nr, nc := r+d.dr, c+d.dc
		if nr < 0 || nr >= n || nc < 0 || nc >= n {
			continue
		}
		vid := cellID(nr, nc, n)
		if (visited>>vid)&1 == 0 {
			est += grid[nr][nc]
		}
	}
	return est
}

func beamSearch(grid [][]int, n, width int) (int, string) {
	startScore := grid[0][0]
	startVisited := uint64(1)
	beam := []*beamNode{{
		priority: -(startScore + heuristic(grid, n, startVisited, 0, 0)),
		depth:    1,
		visited:  startVisited,
		score:    startScore,
	}}

	bestScore := startScore
	bestPath := ""

	for step := 0; step < n*n-1; step++ {
		var candidates []*beamNode
		for _, node := range beam {
			for _, d := range dirs {
				nr, nc := node.r+d.dr, node.c+d.dc
				if nr < 0 || nr >= n || nc < 0 || nc >= n {
					continue
				}
				vid := cellID(nr, nc, n)
				if (node.visited>>vid)&1 != 0 {
					continue
				}
				newVisited := node.visited | (1 << vid)
				newScore := node.score + grid[nr][nc]
				newPath := node.path + string(d.ch)
				h := heuristic(grid, n, newVisited, nr, nc)
				candidates = append(candidates, &beamNode{
					priority: -(newScore + h),
					depth:    node.depth + 1,
					r:        nr,
					c:        nc,
					visited:  newVisited,
					score:    newScore,
					path:     newPath,
				})
				if newScore > bestScore {
					bestScore = newScore
					bestPath = newPath
				}
			}
		}
		if len(candidates) == 0 {
			break
		}
		h := &beamHeap{}
		heap.Init(h)
		limit := width
		if limit > len(candidates) {
			limit = len(candidates)
		}
		for _, c := range candidates {
			heap.Push(h, c)
			if h.Len() > limit {
				heap.Pop(h)
			}
		}
		beam = make([]*beamNode, h.Len())
		for i := len(beam) - 1; i >= 0; i-- {
			beam[i] = heap.Pop(h).(*beamNode)
		}
		_ = step
	}
	return bestScore, bestPath
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	n, grid := readGrid(sc)
	score, path := beamSearch(grid, n, beamWidth)
	fmt.Println(path)
	fmt.Fprintf(os.Stderr, "tiles=%d score=%d beam_width=%d\n", len(path)+1, score, beamWidth)
}

File: python/ch09/beam_fill.py
# python/ch09/beam_fill.py
"""6x6 盤面をビームサーチで埋める（AHC002 簡略版・第9章）."""
from __future__ import annotations

import heapq
import sys
from dataclasses import dataclass, field
from typing import List, Tuple

DIRS = ((-1, 0, "U"), (1, 0, "D"), (0, -1, "L"), (0, 1, "R"))
BEAM_WIDTH = 80


@dataclass(order=True)
class BeamNode:
    neg_score: int
    depth: int
    r: int = field(compare=False)
    c: int = field(compare=False)
    visited: int = field(compare=False)
    score: int = field(compare=False)
    path: str = field(compare=False)


def read_grid(stream) -> Tuple[int, List[List[int]]]:
    lines = stream.read().splitlines()
    n = int(lines[0])
    grid = []
    for i in range(1, n + 1):
        grid.append(list(map(int, lines[i].split())))
    return n, grid


def cell_id(r: int, c: int, n: int) -> int:
    return r * n + c


def heuristic(grid: List[List[int]], n: int, visited: int, r: int, c: int) -> int:
    """未訪問隣接マスの最大値の和（簡易見積もり）."""
    est = 0
    for dr, dc, _ in DIRS:
        nr, nc = r + dr, c + dc
        if 0 <= nr < n and 0 <= nc < n:
            vid = cell_id(nr, nc, n)
            if (visited >> vid) & 1 == 0:
                est += grid[nr][nc]
    return est


def beam_search(grid: List[List[int]], n: int, beam_width: int) -> Tuple[int, str]:
    start_r, start_c = 0, 0
    start_vid = cell_id(start_r, start_c, n)
    start_score = grid[start_r][start_c]
    start_visited = 1 << start_vid

    beam = [
        BeamNode(
            neg_score=-(start_score + heuristic(grid, n, start_visited, start_r, start_c)),
            depth=1,
            r=start_r,
            c=start_c,
            visited=start_visited,
            score=start_score,
            path="",
        )
    ]

    best_score = start_score
    best_path = ""

    for _ in range(n * n - 1):
        candidates: List[BeamNode] = []
        for node in beam:
            for dr, dc, ch in DIRS:
                nr, nc = node.r + dr, node.c + dc
                if not (0 <= nr < n and 0 <= nc < n):
                    continue
                vid = cell_id(nr, nc, n)
                if (node.visited >> vid) & 1:
                    continue
                new_visited = node.visited | (1 << vid)
                new_score = node.score + grid[nr][nc]
                new_path = node.path + ch
                h = heuristic(grid, n, new_visited, nr, nc)
                candidates.append(
                    BeamNode(
                        neg_score=-(new_score + h),
                        depth=node.depth + 1,
                        r=nr,
                        c=nc,
                        visited=new_visited,
                        score=new_score,
                        path=new_path,
                    )
                )
                if new_score > best_score:
                    best_score = new_score
                    best_path = new_path

        if not candidates:
            break
        beam = heapq.nsmallest(beam_width, candidates)

    return best_score, best_path


def main() -> None:
    n, grid = read_grid(sys.stdin)
    score, path = beam_search(grid, n, BEAM_WIDTH)
    print(path)
    print(f"tiles={len(path)+1} score={score} beam_width={BEAM_WIDTH}", file=sys.stderr)


if __name__ == "__main__":
    main()
